### PR TITLE
pgwire: rework DecodeOidDatum to DecodeDatum to parse OidFamily types

### LINKED ISSUE #56112: sentry: *errors.errorString
*safedetails.withSafeDetails: could not find fk for backref %v (1)
structured.go:2934: *withstack.withStack (2)
*assert.withAssertionFailure
(check the extra data payloads)
This issue was autofiled by Sentry. It represents a crash or reported error on a live cluster with telemetry enabled.

Sentry link: https://sentry.io/organizations/cockroach-labs/issues/1989797946/?referrer=webhooks_plugin

Panic message:

>*errors.errorString
*safedetails.withSafeDetails: could not find fk for backref %v (1)
structured.go:2934: *withstack.withStack (2)
*assert.withAssertionFailure
(check the extra data payloads)

<details>
<summary>Stacktrace (expand for inline code snippets):</summary>

https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/sqlbase/structured.go#L2933-L2935 in pkg/sql/sqlbase.(*TableDescriptor).FindFKForBackRef
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/row/cascader.go#L76-L78 in pkg/sql/row.makeDeleteCascader
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/row/deleter.go#L57-L59 in pkg/sql/row.MakeDeleter
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/opt_exec_factory.go#L1596-L1598 in pkg/sql.(*execFactory).ConstructDelete
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/opt/exec/execbuilder/mutation.go#L441-L443 in pkg/sql/opt/exec/execbuilder.(*Builder).buildDelete
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/opt/exec/execbuilder/relational.go#L249-L251 in pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/opt/exec/execbuilder/builder.go#L141-L143 in pkg/sql/opt/exec/execbuilder.(*Builder).build
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/opt/exec/execbuilder/builder.go#L110-L112 in pkg/sql/opt/exec/execbuilder.(*Builder).Build
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/plan_opt.go#L177-L179 in pkg/sql.(*planner).makeOptimizerPlan
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/conn_executor_exec.go#L800-L802 in pkg/sql.(*connExecutor).makeExecPlan
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/conn_executor_exec.go#L690-L692 in pkg/sql.(*connExecutor).dispatchToExecutionEngine
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/conn_executor_exec.go#L480-L482 in pkg/sql.(*connExecutor).execStmtInOpenState
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/conn_executor_exec.go#L95-L97 in pkg/sql.(*connExecutor).execStmt
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/conn_executor.go#L1429-L1431 in pkg/sql.(*connExecutor).execCmd
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/conn_executor.go#L1296-L1298 in pkg/sql.(*connExecutor).run
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/conn_executor.go#L475-L477 in pkg/sql.(*Server).ServeConn
https://github.com/cockroachdb/cockroach/blob/7fd454f880f386cdd0eda6b21b12f6532c14f0db/pkg/sql/pgwire/conn.go#L594-L596 in pkg/sql/pgwire.(*conn).processCommandsAsync.func1
/usr/local/go/src/runtime/asm_amd64.s#L1356-L1358 in runtime.goexit
</details>

```
pkg/sql/sqlbase/structured.go in pkg/sql/sqlbase.(*TableDescriptor).FindFKForBackRef at line 2934
pkg/sql/row/cascader.go in pkg/sql/row.makeDeleteCascader at line 77
pkg/sql/row/deleter.go in pkg/sql/row.MakeDeleter at line 58
pkg/sql/opt_exec_factory.go in pkg/sql.(*execFactory).ConstructDelete at line 1597
pkg/sql/opt/exec/execbuilder/mutation.go in pkg/sql/opt/exec/execbuilder.(*Builder).buildDelete at line 442
pkg/sql/opt/exec/execbuilder/relational.go in pkg/sql/opt/exec/execbuilder.(*Builder).buildRelational at line 250
pkg/sql/opt/exec/execbuilder/builder.go in pkg/sql/opt/exec/execbuilder.(*Builder).build at line 142
pkg/sql/opt/exec/execbuilder/builder.go in pkg/sql/opt/exec/execbuilder.(*Builder).Build at line 111
pkg/sql/plan_opt.go in pkg/sql.(*planner).makeOptimizerPlan at line 178
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).makeExecPlan at line 801
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).dispatchToExecutionEngine at line 691
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execStmtInOpenState at line 481
pkg/sql/conn_executor_exec.go in pkg/sql.(*connExecutor).execStmt at line 96
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).execCmd at line 1430
pkg/sql/conn_executor.go in pkg/sql.(*connExecutor).run at line 1297
pkg/sql/conn_executor.go in pkg/sql.(*Server).ServeConn at line 476
pkg/sql/pgwire/conn.go in pkg/sql/pgwire.(*conn).processCommandsAsync.func1 at line 595
/usr/local/go/src/runtime/asm_amd64.s in runtime.goexit at line 1357
```

| Tag | Value |
|---|---|
| Cockroach Release | `v20.1.3` |
| Cockroach SHA: | 7fd454f880f386cdd0eda6b21b12f6532c14f0db |
| Platform | linux amd64 |
| Distribution | CCL |
| Environment | v20.1.3 |
| Command | start-single-node |
| Go Version | `go1.13.9`|
| # of CPUs |6 |
| # of Goroutines |200 |

### LINKED ISSUE #56193: Can't escape single quote ' in Primary key query
**Describe the problem**

I have a table with a single quote character in the name `'`. I can create the table fine, but cannot retrieve primary key information from the `pg_*` tables.

In running the following query:

```sql
        SELECT a.attname as column_name, format_type(a.atttypid, a.atttypmod) AS data_type
        FROM   pg_index i
        JOIN   pg_attribute a ON a.attrelid = i.indrelid
                            AND a.attnum = ANY(i.indkey)
        WHERE  i.indrelid = 'public.tablewith''char'::regclass
        AND    i.indisprimary;


```
I get the error: `error: lexical error: unterminated string`

This query works fine in Postgres, and I believe I'm escaping the table name correctly?

This is part of a unit test suite for [Beekeeper Studio](https://github.com/beekeeper-studio/beekeeper-studio)

I'm using the latest version of the cockroachdb docker image.

EDIT

I've also tried parameterizing with `$1`, but I get a different error:

```sql
   SELECT a.attname as column_name, format_type(a.atttypid, a.atttypmod) AS data_type
    FROM   pg_index i
    JOIN   pg_attribute a ON a.attrelid = i.indrelid
                        AND a.attnum = ANY(i.indkey)
    WHERE  i.indrelid = $1::regclass
    AND    i.indisprimary;
```

`error: internal error: error in argument for $1: unsupported OID 2205 with format code FormatText`

^ This error happens with ANY table name, regardless of characters in the table name. Again this works fine in Postgres

### PR DESCRIPTION
**Pull request by otan**
Resolves #56193 

pgwire: rework DecodeOidDatum to DecodeDatum to parse OidFamily types

Reworked DecodeOidDatum to DecodeDatum to take in a type, which
encodes additional useful information necessary for ENUMs and oid
family types.

Release note (bug fix): Fixed a bug where reg* types were not parsed
properly over pgwire, COPY or prepared statements.

tree: create ParseDOid method

* Move ParseDOid and associated methods to new function.
* Move datum_test.go to datum_integration_test.go, as it does not import
  datum.go tests.
* Move some of datum_invariants_test.go out into datum_test.go.
* Created new datum_test.go with pure unit tests.

Release note: None

### CODE REVIEW

**Review by rafiss — COMMENTED** [review]
lgtm after fixing the comment. nice refactor ^_^

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/56298#-:-MMHzP_d8VX624xMQ52L:b-r6ybi0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/pgwire/pgwirebase/encoding.go, line 300 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/56298#-MMHx5Tp-naIz1YCiqon:-MMHx5Tp-naIz1YCiqoo:bjvukjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/d9b38b9dbe3201a0c6204cbb92dea966262de6b9/pkg/sql/pgwire/pgwirebase/encoding.go#L300)):*
> ```
> If res is nil
> ```
nit: need to update comment since res is gone

File: pkg/sql/sem/tree/casts_test.go [hunk 1]
```diff
@@ -203,77 +203,3 @@ func TestTupleCastVolatility(t *testing.T) {
 		}
 	}
 }
-
-func TestCastStringToRegClassTableName(t *testing.T) {
```
Comment (rafiss) [inline pkg/sql/sem/tree/casts_test.go:207]: oh just noticed, where did this test go?

Comment (otan) [reply pkg/sql/sem/tree/casts_test.go:207]: pkg/sql/sem/tree/datum_test.go

**Review by rafiss — APPROVED** [review]

**Comment by otan** [comment]
bors r=rafiss

**Pull request merged**